packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -0,0 +1,249 @@
+import { createSnapshot } from '../../client/components/async-local-storage'
+/* eslint-disable import/no-extraneous-dependencies */
+import {
+  renderToReadableStream,
+  decodeReply,
+  createTemporaryReferenceSet as createServerTemporaryReferenceSet,
+} from 'react-server-dom-webpack/server.edge'
+/* eslint-disable import/no-extraneous-dependencies */
+import {
+  createFromReadableStream,
+  encodeReply,
+  createTemporaryReferenceSet as createClientTemporaryReferenceSet,
+} from 'react-server-dom-webpack/client.edge'
+
+import type { StaticGenerationStore } from '../../client/components/static-generation-async-storage.external'
+import { staticGenerationAsyncStorage } from '../../client/components/static-generation-async-storage.external'
+
+type CacheEntry = {
+  value: ReadableStream
+  stale: boolean
+}
+
+interface CacheHandler {
+  get(cacheKey: string | ArrayBuffer): Promise<undefined | CacheEntry>
+  set(cacheKey: string | ArrayBuffer, value: ReadableStream): Promise<void>
+  shouldRevalidateStale: boolean
+}
+
+const cacheHandlerMap: Map<string, CacheHandler> = new Map()
+
+// TODO: Move default implementation to be injectable.
+const defaultCacheStorage: Map<string, ReadableStream> = new Map()
+cacheHandlerMap.set('default', {
+  async get(cacheKey: string | ArrayBuffer) {
+    // TODO: Implement proper caching.
+    if (typeof cacheKey === 'string') {
+      const value = defaultCacheStorage.get(cacheKey)
+      if (value !== undefined) {
+        const [returnStream, newSaved] = value.tee()
+        defaultCacheStorage.set(cacheKey, newSaved)
+        return {
+          value: returnStream,
+          stale: false,
+        }
+      }
+    } else {
+      // TODO: Handle binary keys.
+    }
+    return undefined
+  },
+  async set(cacheKey: string | ArrayBuffer, value: ReadableStream) {
+    // TODO: Implement proper caching.
+    if (typeof cacheKey === 'string') {
+      defaultCacheStorage.set(cacheKey, value)
+    } else {
+      // TODO: Handle binary keys.
+      await value.cancel()
+    }
+  },
+  // In-memory caches are fragile and should not use stale-while-revalidate
+  // semantics on the caches because it's not worth warming up an entry that's
+  // likely going to get evicted before we get to use it anyway.
+  shouldRevalidateStale: false,
+})
+
+const serverManifest: any = null // TODO
+const clientManifest: any = null // TODO
```

Comment:
We could grab this from StaticGenStore maybe if we expose here https://github.com/vercel/next.js/blob/5fd1d537c10e1f63bd693200371dda29a11509d2/packages/next/src/server/app-render/app-render.tsx#L1266 which is attached on `renderOpts`.